clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -142,6 +142,9 @@ public FilterResult filterTo(LogEntryFilter filter, ByteBuffer buffer) {
                     maxTimestamp = shallowRecord.timestamp();
                     shallowOffsetOfMaxTimestamp = shallowEntry.offset();
                 }
+                if (shallowEntry.offset() > maxOffset) {
+                    maxOffset = shallowEntry.offset();
```

Comment:
We need to maintain maxOffset in the else if case when writeOriginalEntry is false, right?